### ISSUE TITLE
Nesting components exhibits odd behavior and rendering glitches

### ISSUE THREAD

**Issue by doughsay** [surface-ui/surface#476]
## Describe the bug

I have a simple utility component that renders one of two slots. You can flip back and forth between which slot is being rendered.  If I nest a copy of this component inside a slot of another copy, things don't work as expected.

## How to reproduce it

I have a minimal repro case here: https://github.com/doughsay/sfc_nested
If you run this phoenix app and go to the `/nested` route you can see it reproduce.

For reference in this issue:

```elixir
# lib/sfc_nested_web/components/either.ex

defmodule SfcNestedWeb.Components.Either do
  use Surface.LiveComponent

  slot(right, required: true)
  slot(left, required: true)

  data(variant, :atom, default: :left)

  @impl true
  def render(assigns) do
    ~F"""
    <div id={@id}>
      <div class="variant">
        {#case @variant}
          {#match :left}
            <strong>Left</strong>
            <a href="#" :on-click="variant" :values={variant: "right"}>
              Right
            </a>
          {#match :right}
            <a href="#" :on-click="variant" :values={variant: "left"}>
              Left
            </a>
            <strong>Right</strong>
        {/case}
      </div>
      <div class="content">
        {#case @variant}
          {#match :left}
            <#slot name="left" />
          {#match :right}
            <#slot name="right" />
        {/case}
      </div>
    </div>
    """
  end

  @impl true
  def handle_event("variant", %{"variant" => variant}, socket) do
    socket = assign(socket, :variant, String.to_existing_atom(variant))

    {:noreply, socket}
  end
end
```

```elixir
# lib/sfc_nested_web/live/nested_live.ex

defmodule SfcNestedWeb.NestedLive do
  use Surface.LiveView

  alias SfcNestedWeb.Components.Either

  @impl true
  def render(assigns) do
    ~F"""
    <Either id="foo">
      <:left>
        Foo Left Content
        <Either id="bar">
          <:left>
            Bar Left Content
          </:left>
          <:right>
            Bar Right content
          </:right>
        </Either>
      </:left>
      <:right>
        Foo Right content
      </:right>
    </Either>
    """
  end
end
```

If you switch from left to right on the `bar` component, it works. If you instead switch to right and back to left on the `foo` component, the bar component breaks, and you can no longer use it. It also appears to think it's in the wrong variant (i.e. it renders "right" in bold, even though it should initialize as "left").

In the minimal repro case I have linked, it doesn't appear to mess up the HTML in any way, it's just a broken state.  But in my application that I discovered this in, the failure is much more severe and actually causes rendering issues. See some screenshots below:

Before:
![DeepinScreenshot_select-area_20210810191935](https://user-images.githubusercontent.com/221693/128959846-13ba6d47-bcbf-4566-a182-df2a5b72c691.png)

After:
![DeepinScreenshot_select-area_20210810191946](https://user-images.githubusercontent.com/221693/128959853-8b2c8656-6acc-4288-8b9f-18058db45a8c.png)

## The behavior you expected

I expect that nesting components should not act differently than if they weren't nested.

## Your Environment
Surface: v0.5.2
LiveView: v0.15.7
Elixir: v1.12.2

**Comment by doughsay** [comment]
More experimenting has revealed that it seems to have something to do with conditionally rendering the slots.  Both slots are marked as required but only one of them is ever rendered at a time.  If instead I use CSS to hide/show the slots, it works as expected, but of course, both slots are present in the DOM.  I need the DOM to only contain one of the two slots at any given time.

I tried to rewrite the slot rendering to use `:if` instead, and it has the same problem:

```html
<div :if={@variant == :left}>
  <#slot name="left" />
</div>
<div :if={@variant == :right}>
  <#slot name="right" />
</div>
```

**Comment by msaraiva** [comment]
Hi @doughsay!

Thanks for the detailed report. I'll take a look at it.

**Comment by msaraiva** [comment]
Hey @doughsay!

It seems the issue is related to LV `0.15`. I updated the dependency to use the latest LV `0.16` version:

```
{:phoenix_live_view, "~> 0.16", override: true}
```

and the problem disappeared. I'm not sure when exactly they fixed the issue nor if they will release a `v0.15.8` patch version with that fix.

I haven't tried using Surface `v0.5.2` with LV `v0.16` extensively as we've been focusing on the upcoming `v0.6` version (which already uses the latest LV version) but its seems it works pretty well as far as I could see.

Feel free to try it out and let me know if you face any other issue. If all works well, we can consider updating Surface's `v0.5` dependencies to also support LV `v0.16` in case the phoenix core team doesn't apply the fix to `v0.15.*`.
